src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs

```diff
@@ -0,0 +1,107 @@
+
+            foreach (var requirement in failure.FailedRequirements)
+            {
+                AppendFailureLine(error, requirement);
+            }
+
+            return error.ToString();
+        }
+
+        protected virtual string GetOperationType(OperationType? operationType)
+        {
+            return operationType switch
+            {
+                OperationType.Query => "query",
+                OperationType.Mutation => "mutation",
+                OperationType.Subscription => "subscription",
+                _ => "operation",
+            };
+        }
+
+        /// <inheritdoc cref="IAuthorizationErrorMessageBuilder"/>
+        public virtual void AppendFailureHeader(StringBuilder error, OperationType? operationType)
+        {
+            error
+                .Append("You are not authorized to run this ")
+                .Append(GetOperationType(operationType))
+                .Append('.');
+        }
+
+        /// <inheritdoc cref="IAuthorizationErrorMessageBuilder"/>
+        public virtual void AppendFailureLine(StringBuilder error, IAuthorizationRequirement authorizationRequirement)
+        {
+            error.AppendLine();
+
+            switch (authorizationRequirement)
+            {
+                case ClaimsAuthorizationRequirement claimsAuthorizationRequirement:
+                    error.Append("Required claim '");
+                    error.Append(claimsAuthorizationRequirement.ClaimType);
+                    if (claimsAuthorizationRequirement.AllowedValues == null || !claimsAuthorizationRequirement.AllowedValues.Any())
+                    {
+                        error.Append("' is not present.");
+                    }
+                    else
+                    {
+                        error.Append("' with any value of '");
+                        error.Append(string.Join(", ", claimsAuthorizationRequirement.AllowedValues));
+                        error.Append("' is not present.");
+                    }
+                    break;
+
+                case DenyAnonymousAuthorizationRequirement _:
+                    error.Append("The current user must be authenticated.");
+                    break;
+
+                case NameAuthorizationRequirement nameAuthorizationRequirement:
+                    error.Append("The current user name must match the name '");
+                    error.Append(nameAuthorizationRequirement.RequiredName);
+                    error.Append("'.");
+                    break;
+
+                case OperationAuthorizationRequirement operationAuthorizationRequirement:
+                    error.Append("Required operation '");
+                    error.Append(operationAuthorizationRequirement.Name);
+                    error.Append("' was not present.");
+                    break;
+
+                case RolesAuthorizationRequirement rolesAuthorizationRequirement:
+                    if (!rolesAuthorizationRequirement.AllowedRoles.Any())
```

Comment:
```suggestion
                    if (rolesAuthorizationRequirement.AllowedRoles == null || !rolesAuthorizationRequirement.AllowedRoles.Any())
```

Why did you change this line?

Reply:
ClaimAuthorizationRequirement.AllowedValues can be null, at least when using .NET Core 3.1. Resharper gave me a warning about that.

Reply:
more correct would be: in .NET Core AllowedValues is not defined as not nullable.

Reply:
> ClaimAuthorizationRequirement.AllowedValues can be null

So... why did you remove check that guards from NRE?
![изображение](https://user-images.githubusercontent.com/21261007/152176735-8a9a4b56-4e7f-4e1b-a0bc-f88da4acad59.png)

Reply:
so the warning is wrong. I didn't know that Any behaves this way. Thank you!